docs/source/format/Columnar.rst

```diff
@@ -208,17 +208,19 @@ right-to-left: ::
               0  0  1  0  1  0  1  1
 
 Arrays having a 0 null count may choose to not allocate the validity
-bitmap. Implementations may choose to always allocate one anyway as a
-matter of convenience, but this should be noted when memory is being
-shared.
+bitmap; how this is represented depends on the implementation (for
+example, a C++ implementation may represent such an "absent" validity
+bitmap using a NULL pointer). Implementations may choose to always allocate
+a validity bitmap anyway as a matter of convenience. Consumers of Arrow
+arrays should be ready to handle those two possibilities.
 
-Nested type arrays except for union types have their own validity bitmap and
-null count regardless of the null count and valid bits of their child arrays.
+Nested type arrays (except for union types as noted above) have their own
+top-level validity bitmap and null count, regardless of the null count and
+valid bits of their child arrays.
 
-Array slots which are null are not required to have a particular
-value; any "masked" memory can have any value and need not be zeroed,
-though implementations frequently choose to zero memory for null
-values.
+Array slots which are null are not required to have a particular value;
+any "masked" memory can have any value and need not be zeroed, though
```

Comment:
it could be worth mentioning that the value in a null slot should be still initialized, i.e. deferencing a pointer pointing to it should not result in undefined behavior.

Reply:
can you clarify what you mean by UB here? Dereferencing a pointer should always be defined given other constraints but tools like ASAN might complain if you try to take actions on the memory location.

Reply:
sorry, I meant to use the defered pointer for anything. I was thinking in terms of leveraging SIMD over null and non-null slots.

An example (in Rust):

```rust
// cargo miri run --example example.rs

// example.rs
fn main() {
    let mut a = Vec::<u8>::with_capacity(4); // similar to aligned_alloc

    let mut dst = a.as_mut_ptr(); // `*mut u8`

    // create a [10, ?, 10, ?]
    for i in 0..4 {
        if i % 2 == 0 {
            unsafe { dst.write(10) }
        } else {
            // let's skip this branch to "save" a `.write`
        }
        dst = unsafe { dst.add(1) };
    }
    unsafe { a.set_len(4) }    // this is unsound in Rust: we can't "expose" uninitialized memory

    let data = a.as_slice();

    // this is UB, we use un-initialized memory; however, we would benefit from being able to perform this op, if e.g.
    // we wanted to perform the +1 over all values in lanes
    data[1] + 1;
}
```

Reply:
This would be a change in specification but a minor one.   I'd point out that i think unless you specify zero initialized you still need to worry about UB because some could choose MAX_INT.    Where adding 1 for the signed values would cause an overflow and UB.    C++ amortized null values by finding chunks of non null sequences 256 values at a time and dispatching to SIMD for these cases

Reply:
In Rust overflowing panics (doesn't UB)

```rust
#![allow(arithmetic_overflow)]

fn main() {
    let a = i32::MAX + 1; // this panics
}
```

but we usually use either saturating or wrapping as we do not want the query engine to panic if the user performs an operation that would panic in a non-null slot.

fwiw `numpy` uses wrapping: `import numpy; a = numpy.array([2147483647], dtype=numpy.int32); print(a + 1) // [-2147483648]`

Reply:
> In Rust overflowing panics in debug and wraps in release (doesn't UB)

OK, there are differences in C++ and Rust's definition of UB so I guess we should be careful with terminology.  I still think this is technically a specification change, and I feel like it has been discussed on the past in the mailing list (but maybe that was zero initialization) and there didn't seem like strong enough consensus to make the change.

Reply:
@jorgecarleitao Sorry, I'm still not entirely sure what initialized means.  Can you give an example of what it would mean for memory to be "initialized" but not "zero initialized"?

Reply:
The rational of this comment is related to general observations about C, C++ and Rust, nicely described e.g. here: https://www.ralfj.de/blog/2019/07/14/uninit.html

all zero-initialized memory is initialized, but not all initialized memory is zero-initialized. My comment/suggestion to require regions to be initialized so that users can access and operate over null slots without triggering undefined behavior.

However, I do agree with Micah that this is beyond this PR.